File: internal/archive/archive.go
package archive

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/iamkoch/go-ralph-go/internal/prd"
)

// Run archives the previous run's prd.json and progress.txt if the PRD branch
// has changed since the last run.
func Run(prdFile, lastBranchFile, progressFile, archiveDir string) {
	if _, err := os.Stat(prdFile); err != nil {
		return
	}
	if _, err := os.Stat(lastBranchFile); err != nil {
		return
	}

	currentBranch := ReadBranch(prdFile)
	lastBranchBytes, err := os.ReadFile(lastBranchFile)
	if err != nil {
		return
	}
	lastBranch := strings.TrimSpace(string(lastBranchBytes))

	if currentBranch == "" || lastBranch == "" || currentBranch == lastBranch {
		return
	}

	date := time.Now().Format("2006-01-02")
	folderName := strings.TrimPrefix(lastBranch, "ralph/")
	archiveFolder := filepath.Join(archiveDir, date+"-"+folderName)

	fmt.Printf("Archiving previous run: %s\n", lastBranch)
	if err := os.MkdirAll(archiveFolder, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "Error creating archive directory: %v\n", err)
		return
	}

	copyFile(prdFile, filepath.Join(archiveFolder, "prd.json"))
	copyFile(progressFile, filepath.Join(archiveFolder, "progress.txt"))
	fmt.Printf("   Archived to: %s\n", archiveFolder)

	InitProgress(progressFile)
}

// ReadBranch reads the branchName field from prd.json.
func ReadBranch(prdFile string) string {
	p, err := prd.Read(prdFile)
	if err != nil {
		return ""
	}
	return p.BranchName
}

// InitProgress creates or resets a progress file with a header.
func InitProgress(path string) {
	content := fmt.Sprintf("# Ralph Progress Log\nStarted: %s\n---\n", time.Now().Format(time.UnixDate))
	os.WriteFile(path, []byte(content), 0644)
}

func copyFile(src, dst string) {
	data, err := os.ReadFile(src)
	if err != nil {
		return
	}
	os.WriteFile(dst, data, 0644)
}

File: src/core/archive.ts
import { mkdir } from "node:fs/promises";
import { join } from "node:path";
import { readPrd } from "./prd.js";

export async function run(
	prdFile: string,
	lastBranchFile: string,
	progressFile: string,
	archiveDir: string,
): Promise<void> {
	const prdExists = await Bun.file(prdFile).exists();
	if (!prdExists) return;

	const lastBranchExists = await Bun.file(lastBranchFile).exists();
	if (!lastBranchExists) return;

	const currentBranch = await readBranch(prdFile);
	const lastBranch = (await Bun.file(lastBranchFile).text()).trim();

	if (!currentBranch || !lastBranch || currentBranch === lastBranch) {
		return;
	}

	const date = new Date().toISOString().slice(0, 10);
	const folderName = lastBranch.replace(/^ralph\//, "");
	const archiveFolder = join(archiveDir, `${date}-${folderName}`);

	console.log(`Archiving previous run: ${lastBranch}`);
	await mkdir(archiveFolder, { recursive: true });

	await copyFile(prdFile, join(archiveFolder, "prd.json"));
	await copyFile(progressFile, join(archiveFolder, "progress.txt"));
	console.log(`   Archived to: ${archiveFolder}`);

	await initProgress(progressFile);
}

export async function readBranch(prdFile: string): Promise<string> {
	try {
		const prd = await readPrd(prdFile);
		return prd.branchName;
	} catch {
		return "";
	}
}

export async function initProgress(path: string): Promise<void> {
	const content = `# Ralph Progress Log\nStarted: ${new Date().toString()}\n---\n`;
	await Bun.write(path, content);
}

async function copyFile(src: string, dst: string): Promise<void> {
	try {
		const data = await Bun.file(src).arrayBuffer();
		await Bun.write(dst, data);
	} catch {
		// Silently ignore copy errors, matching Go behavior
	}
}
